спринт 18.v1

IsaevMilan opened the pull request.

IsaevMilan committed:
Домашка спринт 18.v1

Anna-Malinina commented on app/build.gradle:
```diff
@@ -64,4 +65,8 @@ dependencies {
 
     implementation 'io.insert-koin:koin-android:3.3.0'
+
+    def fragment_version = "1.5.5"
```
Эта переменная нигде не используется, можно убрать

Anna-Malinina commented on app/src/main/java/com/example/myplaylistmaker/app/App.kt:
```diff
@@ -30,7 +31,8 @@ class App : Application() , KoinComponent {
                 PlayerModule,
                 SettingsModule,
-                ViewModelModule
+                ViewModelModule,
+                mediaLibraryModule
```
Какие-то модули у тебя написаны с большой буквы, какие-то с маленькой. Так как это переменные, то они должны быть записаны с маленькой буквы. Будет здорово, если поправишь

IsaevMilan replied:
будет сделано) тороплюсь очень))

Anna-Malinina commented on app/src/main/java/com/example/myplaylistmaker/di/mediaLibraryModule/mediaLibraryModule.kt:
```diff
@@ -0,0 +1,13 @@
+import org.koin.androidx.viewmodel.dsl.viewModel
+import org.koin.dsl.module
+
+val mediaLibraryModule=module {
```
Немного поехало форматирование, равно должно писаться через пробелы. Ctrl + alt + L или cmd + opt + L наше все)

IsaevMilan replied:
спасибо поправлю

Anna-Malinina commented on app/src/main/java/com/example/myplaylistmaker/di/mediaLibraryModule/mediaLibraryModule.kt:
```diff
@@ -0,0 +1,13 @@
+val mediaLibraryModule=module {
+    viewModel { FavoritesViewModel() }
+    viewModel { PlaylistViewModel() }
+    viewModel { MediaLibraryViewModel() }
```
В целом эта вью модель тебе не понадобится, так как никакой логики в активити самой не будет, а в следующем спринте так вообще все на фрагменты переведется

IsaevMilan replied:
ок

Anna-Malinina commented on app/src/main/java/com/example/myplaylistmaker/ui/mediaLibrary/fragments/FavoritesFragment.kt:
```diff
@@ -0,0 +1,31 @@
+
+class FavoritesFragment : Fragment() {
+    private val favoritesViewModel by viewModel<FavoritesViewModel>()
+    private lateinit var nullableFavouritesBinding: FragmentFavoritesBinding
```
Можно убрать nullable из названия, так как биндинг здесь не null, а [lateinit](https://in-kotlin.com/design-patterns/lazy-vs-lateinit-vs-nullable/). И не принято нуллабельность обозначать в название, это всегда должно быть понятно из типа.

Anna-Malinina replied:
Для избежания утечек памяти во фрагменте лучше использовать реально nullable биндинг, а не lateinit: https://developer.android.com/topic/libraries/view-binding#fragments

Anna-Malinina commented on app/src/main/java/com/example/myplaylistmaker/ui/mediaLibrary/fragments/FragmentAdapter.kt:
```diff
@@ -0,0 +1,23 @@
+    fragmentManager: FragmentManager,
+    lifecycle: Lifecycle
+) :
+    FragmentStateAdapter(fragmentManager, lifecycle) {
```
Тоже немного съехала верстка, нужно поднять на 11 строчку

Anna-Malinina commented on app/src/main/java/com/example/myplaylistmaker/ui/mediaLibrary/fragments/SelectPage.kt:
```diff
@@ -0,0 +1,6 @@
+package com.example.myplaylistmaker.ui.mediaLibrary.fragments
+
+
```
Золотой стандарт - одна пустая строка для разделения блоков кода

Anna-Malinina commented on app/src/main/java/com/example/myplaylistmaker/ui/mediaLibrary/fragments/SelectPage.kt:
```diff
@@ -0,0 +1,6 @@
+
+
+interface SelectPage {
+    fun NavigateTo (page:Int)
```
Методы в котлине принято называть [с маленьких букв](https://kotlinlang.org/docs/coding-conventions.html#function-names) (с заглавной буквы их пишут, например, в C#). И немного поехало форматирование, перед скобкой лишняя запятая, а внутри их недостает)

Anna-Malinina commented on app/src/main/java/com/example/myplaylistmaker/ui/mediaLibrary/media_library_activities/MediaLibraryActivity.kt:
```diff
@@ -0,0 +1,60 @@
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        binding = ActivityMediaLibraryBinding.inflate(layoutInflater)
+        binding.arrowBack2.setOnClickListener {
```
А есть arrowBack1?)

Anna-Malinina commented on app/src/main/java/com/example/myplaylistmaker/ui/mediaLibrary/media_library_activities/MediaLibraryActivity.kt:
```diff
@@ -0,0 +1,60 @@
+
+        tabMediator = TabLayoutMediator(binding.tabLayout, binding.pager) { tab, position ->
+            when (position) {
+                0 -> tab.text = "Избранные треки"
```
Строки должны лежать в ресурсах, одна из заповедей андроид разработки) Про переводимость и прочее, думаю, знаешь. На всякий случай напомню, что это все делается для того, чтобы приложение было легко перевести на другой язык. Для разработчиков захардкоженный текст является красным флагом

Anna-Malinina commented on app/src/main/java/com/example/myplaylistmaker/ui/mediaLibrary/media_library_activities/MediaLibraryActivity.kt:
```diff
@@ -0,0 +1,60 @@
+        super.onDestroy()
+        tabMediator.detach()
+    }
+    }
```
Тоже немного съехала верстка

Anna-Malinina commented on app/src/main/java/com/example/myplaylistmaker/ui/mediaLibrary/media_library_activities/MediaLibraryActivity.kt:
```diff
@@ -0,0 +1,60 @@
+    }
+
+
+    override fun NavigateTo(page: Int) {
```
Этот метод есть, но он нигде не используется

Anna-Malinina commented on app/src/main/res/layout/activity_media_library.xml:
```diff
@@ -0,0 +1,55 @@
+
+        <TextView
+            style="@style/TopText"
+            android:layout_marginStart="52dp"
```
Чтобы не позиционировать с помощью марджинов, так как это менее предсказуемое поведение на разных устройствах, нежели позиционирование относительно других элементов, можно использовать Linear Layout, а здесь вполне себе подойдет [тулбар ](https://developer.alexanderklimov.ru/android/views/toolbar.php)

Anna-Malinina commented on app/src/main/res/layout/activity_media_library.xml:
```diff
@@ -0,0 +1,55 @@
+            android:layout_marginTop="0dp"
+            android:layout_marginBottom="0dp"
+            android:fontFamily="@font/ys_display_medium"
+            android:text="@string/ww"
```
Не очень говорящее название для строкового ресурса)

Anna-Malinina commented on app/src/main/res/layout/activity_media_library.xml:
```diff
@@ -0,0 +1,55 @@
+        android:layout_marginTop="16dp"
+        android:textAllCaps="false"
+        app:tabIndicatorColor="?attr/colorButtonNormal"
+        app:tabTextAppearance="@style/Text_14sp"
```
Стили пишут с маленьких букв

Anna-Malinina commented on app/src/main/res/layout/fragment_favorites.xml:
```diff
@@ -0,0 +1,30 @@
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        android:contentDescription="Картинка" />
```
Здесь тоже лучше использовать ресурсы, так как они могут быть переводимыми

Anna-Malinina commented on app/src/main/res/layout/fragment_favorites.xml:
```diff
@@ -0,0 +1,30 @@
+        android:text="@string/emptyMedia"
+        android:fontFamily="@font/ys_display_medium"
+        android:textColor="?attr/colorButtonNormal"
+        android:textSize="@dimen/Font_19sp"
```
И названия ресурсов принято писать с маленькой буквы

IsaevMilan replied:
исправлюсь

Anna-Malinina commented on app/src/main/res/layout/fragment_playlist.xml:
```diff
@@ -0,0 +1,47 @@
+        android:backgroundTint="?attr/colorButtonNormal"
+        android:fontFamily="@font/ys_display_medium"
+        android:text="@string/newPlaylist"
+        android:textAllCaps="false"
+        android:textColor="?attr/colorOnPrimary"
+        android:textSize="14sp"
+        app:cornerRadius="54dp"
```
Кнопки во всем приложение выглядят одинаково, очень хорошо было бы для них создать стиль и использовать его

Anna-Malinina commented on app/src/main/res/layout/fragment_playlist.xml:
```diff
@@ -0,0 +1,47 @@
+        android:gravity="center"
+        android:text="@string/noplaylist"
+        android:textColor="?attr/colorButtonNormal"
+        android:textSize="@dimen/Font_19sp"
```
Текст тоже выглядит одинаково в нескольких местах (как минимум здесь, на экране избранных треков, на экране поиска), это прямое показание к стилю)

Anna-Malinina commented on app/src/main/res/values/strings.xml:
```diff
@@ -46,4 +46,7 @@
     <string name="uncknowncountry">uncknowncountry</string>
     <string name="clearHistoryButton">Очистить историю</string>
+    <string name="emptyMedia">Ваша медиатека пуста</string>
+    <string name="noplaylist">Вы не создали \n ни одного плейлиста</string>
```
Перед и после переноса можно не ставить пробелы, так как, например, на новой строке он рисуется и сдвигает текст. Можно написать `Вы не создали\nни одного плейлиста`

IsaevMilan replied:
отлично спасибо

Anna-Malinina commented on app/src/main/res/values/styles.xml:
```diff
@@ -80,4 +80,25 @@
+        <item name="tint">?attr/color</item>
+    </style>
+
+    <style name="Text_14sp">
```
Название для стиля лучше давать более конкретное, сейчас оно отображает только размер текста, а там ведь еще и шрифт, и цвет. Нужно давать название более говорящее

IsaevMilan replied:
хорошо)

IsaevMilan committed:
Домашка спринт 18.v2

IsaevMilan committed:
Домашка спринт 18.v3
 все критические на мой взгдяд замечания исправлены, но не все рекомендованные исправлю)) пока тороплюсь к дэдлайну)

IsaevMilan committed:
Домашка спринт 18.v4
рекомендейшенс подправил

Anna-Malinina commented on app/src/main/res/values/dimens.xml:
```diff
@@ -1,22 +1,22 @@
-    <dimen name="backArrowMargStart">20dp</dimen>
-    <dimen name="trackTimeMargTop">30dp</dimen>
-    <dimen name="albumNameMargTop">16dp</dimen>
+    <dimen name="textSizePlaylist22">22sp</dimen>
```
Так-с мы немного ушли в недопонимание. Смотри, про конкретику я говорила именно про смысловую, если писать названия с какими-то значениями (числовыми), то это становится жесткой привязкой к значению, и, когда понадобится, например, поменять размер с 22 на 20, то придется название менять. Представляешь сколько будет строк изменений в PR, везде, где это название используется, вместо всего одной строки, где меняется значение. Можешь посмотреть примеры названий на этой странице: https://developer.android.com/guide/topics/resources/more-resources, там нет привязки к значениям в названиях. Прошу прощения, что неясно выразилась(

Anna-Malinina commented on app/src/main/res/values-night-v26/themes.xml:
```diff
@@ -0,0 +1,12 @@
+        <item name="android:background">@color/fullBlack</item>
+        <item name="android:textColor" tools:ignore="DuplicateDefinition">@color/white</item>
+        <item name="android:fontFamily">@font/ys_display_medium</item>
+        <item name="android:textFontWeight" tools:ignore="NewApi">500</item>
```
Не стоит игнорировать NewApi, так как это потенциальный краш приложения с версией ниже 28. У тебя минимальная версия указана 23, то есть краш будет у всех пользователей с 23 по 27 версию. Также вес шрифта уже заложен в самом файле шрифта, нет нужды его еще раз указывать

Anna-Malinina commented on app/src/main/res/values-night-v26/themes.xml:
```diff
@@ -0,0 +1,12 @@
+    <style name="MyToolbar">
+        <item name="theme">@style/ThemeOverlay.AppCompat.Light</item>
+        <item name="android:background">@color/fullBlack</item>
+        <item name="android:textColor" tools:ignore="DuplicateDefinition">@color/white</item>
```
Здесь предупреждения не возникает, поэтому игнор можно убрать

Anna-Malinina commented on app/src/main/res/layout/activity_media_library.xml:
```diff
@@ -4,50 +4,37 @@
-
+        android:layout_height="56dp"
+        android:minHeight="?attr/actionBarSize"
+        app:theme="@style/ThemeOverlay.AppCompat.ActionBar"
```
У тебя тема указана здесь и указана в стиле. Вопрос на засыпку: какая тема применится? Какая тема указана бесполезно и всегда будет игнорироваться?

Ответ [тут](https://habr.com/ru/companies/redmadrobot/articles/461201/#:~:text=%D0%9F%D0%BE%D1%81%D0%BB%D0%B5%D0%B4%D0%BE%D0%B2%D0%B0%D1%82%D0%B5%D0%BB%D1%8C%D0%BD%D0%BE%D1%81%D1%82%D1%8C%20%D0%BF%D1%80%D0%B8%D0%BC%D0%B5%D0%BD%D0%B5%D0%BD%D0%B8%D1%8F%20%D1%82%D0%B5%D0%BC%20%D0%B8%20%D1%81%D1%82%D0%B8%D0%BB%D0%B5%D0%B9%20%D0%BA%D0%BE%20View%2D%D0%BA%D0%BE%D0%BC%D0%BF%D0%BE%D0%BD%D0%B5%D0%BD%D1%82%D1%83)
